prometheus/src/main/scala/filodb/prometheus/ast/Expressions.scala

```diff
@@ -28,57 +28,81 @@ trait Expressions extends Aggregates with Functions {
-
-        case (lh: PeriodicSeries, rh: ScalarExpression) =>
-          val scalar = ScalarFixedDoublePlan(rh.toScalar, RangeParams(timeParams.start, timeParams.step,
-            timeParams.end))
-          val seriesPlan = lh.toSeriesPlan(timeParams)
-          ScalarVectorBinaryOperation(operator.getPlanOperator, scalar, seriesPlan, scalarIsLhs = false)
-
-        case (lh: Function, rh: PeriodicSeries) if lh.isScalarFunction() =>
-          val scalar = lh.toSeriesPlan(timeParams).asInstanceOf[ScalarPlan]
-          val seriesPlanRhs = rh.toSeriesPlan(timeParams)
-          ScalarVectorBinaryOperation(operator.getPlanOperator, scalar, seriesPlanRhs, scalarIsLhs = true)
-
-        case (lh: PeriodicSeries, rh: Function) if rh.isScalarFunction =>
-          val scalar = rh.toSeriesPlan(timeParams).asInstanceOf[ScalarPlan]
-          val seriesPlanlhs = lh.toSeriesPlan(timeParams)
-          ScalarVectorBinaryOperation(operator.getPlanOperator, scalar, seriesPlanlhs, scalarIsLhs = false)
-
-        case (lh: PeriodicSeries, rh: PeriodicSeries) =>
-          val seriesPlanLhs = lh.toSeriesPlan(timeParams)
-          val seriesPlanRhs = rh.toSeriesPlan(timeParams)
-          val cardinality = if (operator.getPlanOperator.isInstanceOf[SetOperator])
-            Cardinality.ManyToMany
-          else
-            vectorMatch.map(_.cardinality.cardinality).getOrElse(Cardinality.OneToOne)
-
-          val matcher = vectorMatch.flatMap(_.matching)
-          val onLabels = matcher.filter(_.isInstanceOf[On]).map(_.labels)
-          val ignoringLabels = matcher.filter(_.isInstanceOf[Ignoring]).map(_.labels)
-
-          BinaryJoin(seriesPlanLhs, operator.getPlanOperator, cardinality, seriesPlanRhs,
-            onLabels.getOrElse(Nil), ignoringLabels.getOrElse(Nil),
-            vectorMatch.flatMap(_.grouping).map(_.labels).getOrElse(Nil))
-
-        case _ => throw new UnsupportedOperationException("Invalid operands")
+      if (hasScalarResult(lhs) && hasScalarResult(rhs)) {
+        val rangeParams = RangeParams(timeParams.start, timeParams.step, timeParams.end)
+
+        (lhs, rhs) match {
+          // 3 + 4
+          case (lh: ScalarExpression, rh: ScalarExpression) =>
+            ScalarBinaryOperation(operator.getPlanOperator, Left(lh.toScalar), Left(rh.toScalar), rangeParams)
+          // (2 + 3) + 5
+          case (lh: BinaryExpression, rh: ScalarExpression) => ScalarBinaryOperation(operator.getPlanOperator,
+            Right(lh.toSeriesPlan(timeParams).asInstanceOf[ScalarBinaryOperation]), Left(rh.toScalar), rangeParams)
+          // (2 + 3) + (5 - 6)
+          case (lh: BinaryExpression, rh: BinaryExpression) => ScalarBinaryOperation(operator.getPlanOperator,
+            Right(lh.toSeriesPlan(timeParams).asInstanceOf[ScalarBinaryOperation]),
+            Right(rh.toSeriesPlan(timeParams).asInstanceOf[ScalarBinaryOperation]), rangeParams)
+        }
+      } else {
+
+        (lhs, rhs) match {
```

Comment:
Could simplify the nested pattern matching if possible. Both are `(lhs, rhs) match`

Reply:
For ScalarBinaryOperation we need to check that lhs and rhs yield fixed scalar/double result. BinaryExpression is also type of  PeriodicSeries so it can overlap with other cases like 
`case (lh: PeriodicSeries, rh: PeriodicSeries)`